### ISSUE TITLE
Should allow pressure-vessel to make /snap, /var/lib/snapd/hostfs visible to its container

### ISSUE THREAD

**Issue by smcv** [canonical/steam-snap#359]
### Current Behavior

When pressure-vessel creates a container environment, we are currently unable to provide the correct `/usr/share/libdrm`, `/usr/share/drirc.d` and `/usr/share/nvidia` to games inside the container.

I have been told by Nvidia driver maintainers that access to the files shipped in `/usr/share/nvidia` is functionally necessary for some driver features. Similarly, I've been told by Mesa maintainers that `/usr/share/libdrm` and `/usr/share/drirc.d` are functionally necessary for some games.

You can tell this is happening, and you can tell it's a problem, because these warnings are logged:

```
pressure-vessel-wrap[30807]: W: "snap/steam/171/graphics/usr/share/libdrm" is unlikely to appear in "/run/host"
pressure-vessel-wrap[30807]: W: "snap/steam/171/graphics/usr/share/libdrm" is unlikely to appear in "/run/host"
pressure-vessel-wrap[30807]: W: "var/lib/snapd/hostfs/usr/share/drirc.d" is unlikely to appear in "/run/host"
pressure-vessel-wrap[30807]: W: "snap/steam/171/graphics/usr/share/drirc.d" is unlikely to appear in "/run/host"
pressure-vessel-wrap[30807]: W: "var/lib/snapd/hostfs/usr/share/nvidia" is unlikely to appear in "/run/host"
pressure-vessel-wrap[30807]: W: "var/lib/snapd/hostfs/usr/share/nvidia" is unlikely to appear in "/run/host"
```

This is partly pressure-vessel's fault, and to solve it, I want to apply https://gitlab.steamos.cloud/steamrt/steam-runtime-tools/-/merge_requests/556. Unfortunately, if we do that, the container won't start at all, with an error message like this logged:

`bwrap: Can't bind mount /oldroot/var/lib/snapd/hostfs on /newroot/var/lib/snapd/hostfs: Unable to mount source on destination: Permission denied`

This is because snapd's steam_support AppArmor profile doesn't allow what we would like to do.

### Expected Behavior

Should allow mounting `/oldroot/snap` on `/newroot/snap`, so that pressure-vessel can do its part of solving this. The ability to make `/snap/` inside our container be identical to `/snap/` outside our container does not seem like a security hazard!

Similarly, should allow `/var/lib/snapd/hostfs`.

Should also allow mounting individual files within `/var/lib/snapd/hostfs/usr/share/`, not just `/var/lib/snapd/hostfs/usr/lib/@{multiarch}/`.

When using a suitable version of pressure-vessel, the files `/usr/share/libdrm/amdgpu.ids`, `/usr/share/drirc.d/00-mesa-defaults.conf` and `/usr/share/nvidia/nvoptix.bin` should all be accessible from within the pressure-vessel container (for example if you set a game's launch options to `PRESSURE_VESSEL_SHELL=instead %command%` and use the resulting `xterm`).

### Steps To Reproduce

* Run /snap/bin/steam in a terminal
* Run a Proton or Steam Linux Runtime game
* Observe warnings

and

* Run /snap/bin/steam in a terminal
* Replace `SteamLinuxRuntime_soldier/pressure-vessel` and `SteamLinuxRuntime_sniper/pressure-vessel` with the one from https://gitlab.steamos.cloud/steamrt/steam-runtime-tools/-/jobs/429313/artifacts/raw/_build/production/pressure-vessel-bin+src.tar.gz
* Run a Proton or Steam Linux Runtime game
* Observe game not starting at all, and error message from bwrap in the terminal output

### Workarounds

I was able to make this mostly work by hacking these into `/var/lib/snapd/apparmor/profiles/snap.steam.steam`:

```
mount options=(rw, rbind) /oldroot/var/lib/snapd/hostfs/usr/share/** -> /newroot/**,
mount options=(rw, rbind) /oldroot/var/lib/snapd/hostfs/ -> /newroot/var/lib/snapd/hostfs/,
mount options=(rw, rbind) /oldroot/snap/ -> /newroot/snap/,
# from https://github.com/snapcore/snapd/pull/13489
mount fstype=tmpfs options=(rw, nosuid, nodev) tmpfs -> /newroot/**,
```

and reloading the profile, but I think actually being able to read the Nvidia files will also require something like:

```
/var/lib/snapd/hostfs/usr/share/nvidia/** rm,
```

(untested).

**Comment by smcv** [comment]
The latest pressure-vessel build https://repo.steampowered.com/pressure-vessel/snapshots/0.20240321.0/ has a mechanism to expose these directories to games in the container. This is likely to be in the next sniper beta (hopefully later today, but I don't control the exact schedule) and also in a future Steam beta.

You can try it early by unpacking https://repo.steampowered.com/pressure-vessel/snapshots/0.20240321.0/pressure-vessel-bin.tar.gz into (for example) `~/tmp/pressure-vessel`, and then running Steam with `PRESSURE_VESSEL_PREFIX=$HOME/tmp/pressure-vessel` in the environment.

Because we can't yet count on having an updated version of Snap, this mechanism is disabled by default for now as a workaround for this issue. You can activate it for testing by adding environment variable `PRESSURE_VESSEL_WORKAROUNDS="-steam-snap#359"` (which will cause pressure-vessel to fail with current versions of snapd, but will hopefully work with the snapd where https://github.com/snapcore/snapd/pull/13633 has been integrated).

It's a space- or comma-separated list, and each item starts with a `+` or `-` to activate or deactivate a workaround, so to change the state of more than one workaround you can say something like `PRESSURE_VESSEL_WORKAROUNDS="-steam-snap#359 +steam-snap#356"`.

**Comment by smcv** [comment]
If I'm keeping track correctly, this was fixed in snapd 2.62, which is mandatory since #367. As a result, the pressure-vessel version included in yesterday's SLR soldier and sniper beta releases disables our workaround for this issue by default. If there are no regression reports in the next few days/weeks, we will remove the workaround completely.

If it becomes necessary to re-enable the workaround, you can do this:

1. Report an issue to https://github.com/ValveSoftware/steam-runtime/issues so we can enable it by default again, and so that we know that we can't delete the implementation of the workaround yet
2. Run with `PRESSURE_VESSEL_WORKAROUNDS="+steam-snap#359"` in the environment (if you need to use more than one workaround, it's a space-separated list)

**Comment by ashuntu** [comment]
Closing as it seems resolved.
